eth/watchers/senderwatcher.go

```diff
@@ -0,0 +1,218 @@
+	sub := sw.watcher.Subscribe(events)
+	defer sub.Unsubscribe()
+	for {
+		select {
+		case <-sw.quit:
+			return
+		case err := <-sub.Err():
+			glog.Error(err)
+		case events := <-events:
+			sw.handleBlockEvents(events)
+		}
+	}
+}
+
+// Stop watching for events
+func (sw *SenderWatcher) Stop() {
+	close(sw.quit)
+}
+
+// Clear removes a key-value pair from the map
+func (sw *SenderWatcher) Clear(addr ethcommon.Address) {
+	sw.mu.Lock()
+	defer sw.mu.Unlock()
+	if _, ok := sw.senders[addr]; ok {
+		delete(sw.senders, addr)
+	}
+}
+
+func (sw *SenderWatcher) handleBlockEvents(events []*blockwatch.Event) {
+	for _, event := range events {
+		for _, log := range event.BlockHeader.Logs {
+			if event.Type == blockwatch.Removed {
+				log.Removed = true
+			}
+			if err := sw.handleLog(log); err != nil {
+				glog.Error(err)
+			}
+		}
+	}
+}
+
+func (sw *SenderWatcher) handleLog(log types.Log) error {
+	eventName, err := sw.dec.FindEventName(log)
+	if err != nil {
+		// Noop if we cannot find the event name
+		return nil
+	}
+
+	sw.mu.Lock()
+	defer sw.mu.Unlock()
+
+	var sender ethcommon.Address
+	switch eventName {
+	case "DepositFunded":
+		var depositFunded contracts.TicketBrokerDepositFunded
+		if err := sw.dec.Decode("DepositFunded", log, &depositFunded); err != nil {
+			return fmt.Errorf("failed to decode DepositFunded event: %v", err)
+		}
+		sender = depositFunded.Sender
+		if info, ok := sw.senders[sender]; ok && !log.Removed {
+			info.Deposit.Add(info.Deposit, depositFunded.Amount)
+		}
+	case "ReserveFunded":
+		var reserveFunded contracts.TicketBrokerReserveFunded
+		if err := sw.dec.Decode("ReserveFunded", log, &reserveFunded); err != nil {
+			return fmt.Errorf("failed to decode ReserveFunded event: %v", err)
+		}
+		sender = reserveFunded.ReserveHolder
+		if info, ok := sw.senders[sender]; ok && !log.Removed {
+			info.Reserve.Add(info.Reserve, reserveFunded.Amount)
+		}
+	case "Withdrawal":
+		var withdrawal contracts.TicketBrokerWithdrawal
+		if err := sw.dec.Decode("Withdrawal", log, &withdrawal); err != nil {
+			return fmt.Errorf("failed to decode Withdrawal event: %v", err)
+		}
+		sender = withdrawal.Sender
+		if info, ok := sw.senders[sender]; ok && !log.Removed {
+			info.Deposit = big.NewInt(0)
+			info.Reserve = big.NewInt(0)
+		}
+	case "WinningTicketTransfer":
+		var winningTicketTransfer contracts.TicketBrokerWinningTicketTransfer
+		if err := sw.dec.Decode("WinningTicketTransfer", log, &winningTicketTransfer); err != nil {
+			return fmt.Errorf("failed to decode WinningTicketTransfer event: %v", err)
+		}
+		amount := winningTicketTransfer.Amount
+		sender = winningTicketTransfer.Sender
+
+		if info, ok := sw.senders[sender]; ok && !log.Removed {
+			// See if amount > deposit
+			if info.Deposit.Cmp(amount) < 0 {
+				// Draw from reserve
+				// ReserveFrozen will be handled in it's own event log handler
+				diff := new(big.Int).Abs(new(big.Int).Sub(info.Deposit, amount))
```

Comment:
I don't think we need to take the absolute value here since we know that `info.Deposit` is less than `amount`. We could just write `diff := new(big.Int).Sub(amount, info.Deposit)`

Reply:
fixed in 4530eff